# Análisis Exploratorio de Datos (EDA) con NumPy y Pandas sobre World Happiness
Aplicar el ciclo completo de exploración de datos: reconocimiento inicial, limpieza (cleaning), análisis descriptivo y visualización, usando un dataset real de Kaggle


In [ ]:
import pandas as pd
import numpy as np
import matplotlib as plt

In [ ]:
# Lectura de data set e inicialización de cada uno para exploración respecto a su año
first_year=2015
last_year=2019

dfs_initial = []

for year in range(2015, 2020):
    df = pd.read_csv(f"./data/{year}.csv")
    df["year"] = year
    dfs_initial.append(df)

In [ ]:
dfs_initial[0].head() # Data frame del año 2015

,Country,Region,Happiness Rank,Happiness Score,Standard Error,Economy (GDP per Capita),Family,Health (Life Expectancy),Freedom,Trust (Government Corruption),Generosity,Dystopia Residual,year
0,Switzerland,Western Europe,1,7.587,0.03411,1.39651,1.34951,0.94143,0.66557,0.41978,0.29678,2.51738,2015
1,Iceland,Western Europe,2,7.561,0.04884,1.30232,1.40223,0.94784,0.62877,0.14145,0.43630,2.70201,2015
2,Denmark,Western Europe,3,7.527,0.03328,1.32548,1.36058,0.87464,0.64938,0.48357,0.34139,2.49204,2015
3,Norway,Western Europe,4,7.522,0.03880,1.45900,1.33095,0.88521,0.66973,0.36503,0.34699,2.46531,2015
4,Canada,North America,5,7.427,0.03553,1.32629,1.32261,0.90563,0.63297,0.32957,0.45811,2.45176,2015


In [ ]:
dfs_initial[1].head() # Data frame del año 2016

,Country,Region,Happiness Rank,Happiness Score,Lower Confidence Interval,Upper Confidence Interval,Economy (GDP per Capita),Family,Health (Life Expectancy),Freedom,Trust (Government Corruption),Generosity,Dystopia Residual,year
0,Denmark,Western Europe,1,7.526,7.460,7.592,1.44178,1.16374,0.79504,0.57941,0.44453,0.36171,2.73939,2016
1,Switzerland,Western Europe,2,7.509,7.428,7.590,1.52733,1.14524,0.86303,0.58557,0.41203,0.28083,2.69463,2016
2,Iceland,Western Europe,3,7.501,7.333,7.669,1.42666,1.18326,0.86733,0.56624,0.14975,0.47678,2.83137,2016
3,Norway,Western Europe,4,7.498,7.421,7.575,1.57744,1.12690,0.79579,0.59609,0.35776,0.37895,2.66465,2016
4,Finland,Western Europe,5,7.413,7.351,7.475,1.40598,1.13464,0.81091,0.57104,0.41004,0.25492,2.82596,2016


In [ ]:
dfs_initial[2].head() # Data frame del año 2017

,Country,Happiness.Rank,Happiness.Score,Whisker.high,Whisker.low,Economy..GDP.per.Capita.,Family,Health..Life.Expectancy.,Freedom,Generosity,Trust..Government.Corruption.,Dystopia.Residual,year
0,Norway,1,7.537,7.594445,7.479556,1.616463,1.533524,0.796667,0.635423,0.362012,0.315964,2.277027,2017
1,Denmark,2,7.522,7.581728,7.462272,1.482383,1.551122,0.792566,0.626007,0.355280,0.400770,2.313707,2017
2,Iceland,3,7.504,7.622030,7.385970,1.480633,1.610574,0.833552,0.627163,0.475540,0.153527,2.322715,2017
3,Switzerland,4,7.494,7.561772,7.426227,1.564980,1.516912,0.858131,0.620071,0.290549,0.367007,2.276716,2017
4,Finland,5,7.469,7.527542,7.410458,1.443572,1.540247,0.809158,0.617951,0.245483,0.382612,2.430182,2017


In [ ]:
dfs_initial[3].head() # Data frame del año 2018

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption,year
0,1,Finland,7.632,1.305,1.592,0.874,0.681,0.202,0.393,2018
1,2,Norway,7.594,1.456,1.582,0.861,0.686,0.286,0.340,2018
2,3,Denmark,7.555,1.351,1.590,0.868,0.683,0.284,0.408,2018
3,4,Iceland,7.495,1.343,1.644,0.914,0.677,0.353,0.138,2018
4,5,Switzerland,7.487,1.420,1.549,0.927,0.660,0.256,0.357,2018


In [ ]:
dfs_initial[4].head() # Data frame del año 2019

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption,year
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393,2019
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410,2019
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341,2019
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118,2019
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298,2019


Los cinco archivos usan esquemas distintos: nombres diferentes para la misma variable como: `Country o region, GDP per capital, Perceptions of corruption, Freedom, family`, las cuales afectan directamente el score de felicidad segun el data set; así mismo hay columnas que solo aparecen en algunos años como: `Whisker.low o Whisker.high`

In [74]:
# Concatenacion temporal de los dataframes
df_raw = pd.concat(dfs_initial, ignore_index=True)

print("shape:", df_raw.shape)
df_raw.info()

shape: (782, 31)
<class 'pandas.DataFrame'>
RangeIndex: 782 entries, 0 to 781
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        470 non-null    str    
 1   Region                         315 non-null    str    
 2   Happiness Rank                 315 non-null    float64
 3   Happiness Score                315 non-null    float64
 4   Standard Error                 158 non-null    float64
 5   Economy (GDP per Capita)       315 non-null    float64
 6   Family                         470 non-null    float64
 7   Health (Life Expectancy)       315 non-null    float64
 8   Freedom                        470 non-null    float64
 9   Trust (Government Corruption)  315 non-null    float64
 10  Generosity                     782 non-null    float64
 11  Dystopia Residual              315 non-null    float64
 12  year                           782 non-null 

En un primer momento hicimos una concatenación directa, pero pandas alinea por nombre de columna y el resultado fue una data frame con columnas repetidas pero con nombres distintos, saliendo una columna de **782 filas x 31 columnas** y casi todas incompletas. `Generosity` es la única con 782 valores no nulos, porque es la única columna que se llama igual en los cinco archivos. Esto generaba gran cantidad de valores nulos, por lo cual fue necesario estandarizarlas. 

### Estandarización y limpieza de Dataset 

Primero identificamos cuales serian las columnas a estandarizar, para ello analizamos el dataset encontrando que columnas como: `Family y Social Support, Trust (Government Corruption) y Perceptions of corruption` representaban lo mismo, así que se creo un diccionario con las columnas a renombrar por año.

In [ ]:
RENAME_2015_2016 = {
    "Happiness Rank": "Overall rank",
    "Happiness Score": "Score",
    "Economy (GDP per Capita)": "GDP per Capita",
    "Family": "Social support",
    "Trust (Government Corruption)": "Perceptions of corruption",
}

RENAME_2017 = {
    "Happiness.Rank": "Overall rank",
    "Happiness.Score": "Score",
    "Economy..GDP.per.Capita.": "GDP per Capita",
    "Family": "Social support",
    "Health..Life.Expectancy.": "Health (Life Expectancy)",
    "Trust..Government.Corruption.": "Perceptions of corruption",
}

RENAME_2018_2019 = {
    "Country or region": "Country",
    "GDP per capita": "GDP per Capita",
    "Healthy life expectancy": "Health (Life Expectancy)",
    "Freedom to make life choices": "Freedom",
}

RENAME_BY_YEAR = {
    2015: RENAME_2015_2016,
    2016: RENAME_2015_2016,
    2017: RENAME_2017,
    2018: RENAME_2018_2019,
    2019: RENAME_2018_2019,
}

# rename() devuelve una copia, asi que dfs_initial queda intacto
dfs = []
for df in dfs_initial:
    year = df["year"].iloc[0]
    dfs.append(df.rename(columns=RENAME_BY_YEAR[year]))


: 

#### Normalizar los nombres de país

`Region` solo existe en 2015 y 2016, así que la vamos a propagar cruzando por el nombre
del país. Pero antes hay que arreglar un detalle: varios países cambiaron de nombre entre
ediciones del reporte. Si cruzamos sin normalizar, esos países quedan sin región y parece
un bug del script cuando en realidad es un problema de datos.

In [76]:
# Paises de 2017-2019 que NO aparecen con ese nombre en 2015/2016
conocidos = set(dfs[0]["Country"]) | set(dfs[1]["Country"])

for df in dfs[2:]:
    year = df["year"].iloc[0]
    sin_match = sorted(set(df["Country"]) - conocidos)
    print(f"{year}: {len(sin_match)} sin match -> {sin_match}")

2017: 2 sin match -> ['Hong Kong S.A.R., China', 'Taiwan Province of China']
2018: 2 sin match -> ['Northern Cyprus', 'Trinidad & Tobago']
2019: 4 sin match -> ['Gambia', 'North Macedonia', 'Northern Cyprus', 'Trinidad & Tobago']


Se identifican 8 casos que es necesario normalizar, así que incluimos nuevamente un diccionario que remplace los nombres, asi como corregir `Somaliland region` (2015) vs `Somaliland Region` (2016), que es una
inconsistencia dentro de los propios archivos de referencia.

In [77]:
COUNTRY_RENAMES = {
    "Hong Kong S.A.R., China": "Hong Kong",
    "Taiwan Province of China": "Taiwan",
    "Northern Cyprus": "North Cyprus",
    "Trinidad & Tobago": "Trinidad and Tobago",
    "North Macedonia": "Macedonia",
    "Somaliland region": "Somaliland Region",
}

for df in dfs:
    df["Country"] = df["Country"].replace(COUNTRY_RENAMES)



### Propagar región

Decidimos conservar la region, puesto que nos sirve para la realización una analisis y no solo tener los datos de un país, por ello fue necesario propagar la columna de region a los datasets de 2017,2018,2019 a partir de la construcción de un diccionario pais -> region obtenida de 2015 y 2016, adiccional agregamos Gambia a mano ya que en estos dataset no se encuentra su región

In [78]:
con_region = (
    pd.concat([dfs[0], dfs[1]])[["Country", "Region"]]
    .drop_duplicates("Country")
)
region_map = dict(zip(con_region["Country"], con_region["Region"]))

# Gambia solo aparece en 2019, no hay de donde heredar la region
region_map["Gambia"] = "Sub-Saharan Africa"

print("paises en el mapa:", len(region_map))

for df in dfs:
    df["Region"] = df["Country"].map(region_map)
    print(df["year"].iloc[0], "- regiones nulas:", df["Region"].isna().sum())

paises en el mapa: 164
2015 - regiones nulas: 0
2016 - regiones nulas: 0
2017 - regiones nulas: 0
2018 - regiones nulas: 0
2019 - regiones nulas: 0


#### Eliminar las columnas fuera del esquema

Se van `Standard Error` (2015), `Lower/Upper Confidence Interval` (2016),
`Whisker.high` / `Whisker.low` (2017) y `Dystopia Residual` (2015-2017). Seleccionar el
esquema de una vez elimina lo sobrante y deja las columnas en el mismo orden en los cinco
dataframes.

In [ ]:
SCHEMA = [
    "year",
    "Overall rank",
    "Country",
    "Region",
    "Score",
    "GDP per Capita",
    "Social support",
    "Health (Life Expectancy)",
    "Freedom",
    "Generosity",
    "Perceptions of corruption",
]

dfs = [df[SCHEMA] for df in dfs]

2015 (158, 11)
2016 (157, 11)
2017 (155, 11)
2018 (156, 11)
2019 (156, 11)


#### Concatenar

Ahora sí: los cinco dataframes tienen exactamente las mismas columnas, así que el concat apila filas sin generar nulos.

In [ ]:
df = pd.concat(dfs, ignore_index=True)
df

,year,Overall rank,Country,Region,Score,GDP per Capita,Social support,Health (Life Expectancy),Freedom,Generosity,Perceptions of corruption
0,2015,1,Switzerland,Western Europe,7.587,1.39651,1.34951,0.94143,0.66557,0.29678,0.41978
1,2015,2,Iceland,Western Europe,7.561,1.30232,1.40223,0.94784,0.62877,0.43630,0.14145
2,2015,3,Denmark,Western Europe,7.527,1.32548,1.36058,0.87464,0.64938,0.34139,0.48357
3,2015,4,Norway,Western Europe,7.522,1.45900,1.33095,0.88521,0.66973,0.34699,0.36503
4,2015,5,Canada,North America,7.427,1.32629,1.32261,0.90563,0.63297,0.45811,0.32957


In [83]:
df.tail()

,year,Overall rank,Country,Region,Score,GDP per Capita,Social support,Health (Life Expectancy),Freedom,Generosity,Perceptions of corruption
777,2019,152,Rwanda,Sub-Saharan Africa,3.334,0.359,0.711,0.614,0.555,0.217,0.411
778,2019,153,Tanzania,Sub-Saharan Africa,3.231,0.476,0.885,0.499,0.417,0.276,0.147
779,2019,154,Afghanistan,Southern Asia,3.203,0.350,0.517,0.361,0.000,0.158,0.025
780,2019,155,Central African Republic,Sub-Saharan Africa,3.083,0.026,0.000,0.105,0.225,0.235,0.035
781,2019,156,South Sudan,Sub-Saharan Africa,2.853,0.306,0.575,0.295,0.010,0.202,0.091


El resultado final es un dataframe limpio de **782 filas x 10 columnas**, a diferencia del otro que tenia 31 columnas y muchas nulas.

In [85]:
df.shape

(782, 11)

Las columnas de texto son Country y Region, y las demas son de tipo numerico.

In [81]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 782 entries, 0 to 781
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   year                       782 non-null    int64  
 1   Overall rank               782 non-null    int64  
 2   Country                    782 non-null    str    
 3   Region                     782 non-null    str    
 4   Score                      782 non-null    float64
 5   GDP per Capita             782 non-null    float64
 6   Social support             782 non-null    float64
 7   Health (Life Expectancy)   782 non-null    float64
 8   Freedom                    782 non-null    float64
 9   Generosity                 782 non-null    float64
 10  Perceptions of corruption  781 non-null    float64
dtypes: float64(7), int64(2), str(2)
memory usage: 67.3 KB


Score tiene una desviacion estandar muy pequeña, lo que significa que los puntajes tienen una variacion muy pequeña con respecto a la media. Esto puede ser un indicio de que los puntajes estan muy concentrados alrededor de la media, lo que puede dificultar la diferenciacion entre los paises en terminos de felicidad.

En la percepcion de corrupcion hasta el 75% de los datos era de 0.156, pero el dato maximo es de 0.55 lo que puede indicar que este dato es un outlier y encuestas indican que la corrupcion no es tan grave o es un dato invalido

En todas las categorias que influyen en el score, los minimos se encuentran en 0, lo que indica que puedo haber un error en este registro con un dato invalido



In [86]:
df.describe()

,year,Overall rank,Score,GDP per Capita,Social support,Health (Life Expectancy),Freedom,Generosity,Perceptions of corruption
count,782.000000,782.000000,782.000000,782.000000,782.000000,782.000000,782.000000,782.000000,781.000000
mean,2016.993606,78.698210,5.379018,0.916047,1.078392,0.612416,0.411091,0.218576,0.125436
std,1.417364,45.182384,1.127456,0.407340,0.329548,0.248309,0.152880,0.122321,0.105816
min,2015.000000,1.000000,2.693000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2016.000000,40.000000,4.509750,0.606500,0.869363,0.440183,0.309768,0.130000,0.054000
50%,2017.000000,79.000000,5.322000,0.982205,1.124735,0.647310,0.431000,0.201982,0.091000
75%,2018.000000,118.000000,6.189500,1.236187,1.327250,0.808000,0.531000,0.278832,0.156030
max,2019.000000,158.000000,7.769000,2.096000,1.644000,1.141000,0.724000,0.838075,0.551910


In [89]:
corruption= df["Perceptions of corruption"].to_numpy() 

crrpt_mean = np.mean(corruption)
crrpt_median = np.median(corruption)
crrpt_std = np.std(corruption)


(1*100) / 782 = 0.1278% 

In [92]:
df.isnull().sum()



year                         0
Overall rank                 0
Country                      0
Region                       0
Score                        0
GDP per Capita               0
Social support               0
Health (Life Expectancy)     0
Freedom                      0
Generosity                   0
Perceptions of corruption    1
dtype: int64

No hay nulos en el dataset, lo que indica que la limpieza de datos fue exitosa y que los datos estan completos y listos para su analisis.
No hay duplicados
Ya se estandarizaron los nombres de paises y regiones, lo que permite un analisis mas preciso y confiable.


In [ ]:
df.duplicated()

,year,Overall rank,Country,Region,Score,GDP per Capita,Social support,Health (Life Expectancy),Freedom,Generosity,Perceptions of corruption
0,2015,1,Switzerland,Western Europe,7.587,1.39651,1.34951,0.94143,0.66557,0.29678,0.41978
1,2015,2,Iceland,Western Europe,7.561,1.30232,1.40223,0.94784,0.62877,0.43630,0.14145
2,2015,3,Denmark,Western Europe,7.527,1.32548,1.36058,0.87464,0.64938,0.34139,0.48357
3,2015,4,Norway,Western Europe,7.522,1.45900,1.33095,0.88521,0.66973,0.34699,0.36503
4,2015,5,Canada,North America,7.427,1.32629,1.32261,0.90563,0.63297,0.45811,0.32957
...,...,...,...,...,...,...,...,...,...,...,...
777,2019,152,Rwanda,Sub-Saharan Africa,3.334,0.35900,0.71100,0.61400,0.55500,0.21700,0.41100
778,2019,153,Tanzania,Sub-Saharan Africa,3.231,0.47600,0.88500,0.49900,0.41700,0.27600,0.14700
779,2019,154,Afghanistan,Southern Asia,3.203,0.35000,0.51700,0.36100,0.00000,0.15800,0.02500
780,2019,155,Central African Republic,Sub-Saharan Africa,3.083,0.02600,0.00000,0.10500,0.22500,0.23500,0.03500


In [101]:
df['Country'].nunique()

164

Punto 11
variable a analizar percetage of corruption

Q1 = 0.054

Q2 = 0.091 

Q3 = 0.156

IQR = Q3 - Q1 = 0.156 - 0.054 = 0.102

lower_bound = Q1 - 1.5 * IQR = 0.054 - 1.5 * 0.102 = -0.099
upper_bound = Q3 + 1.5 * IQR = 0.156 + 1.5 * 0.102 = 0.309


In [104]:
Q1 = 0.054

Q2 = 0.091 

Q3 = 0.156

IQR = 0.102

lower_bound = -0.099
upper_bound = 0.309

df[(df["Perceptions of corruption"] < lower_bound) ]

,year,Overall rank,Country,Region,Score,GDP per Capita,Social support,Health (Life Expectancy),Freedom,Generosity,Perceptions of corruption


In [105]:
df[(df["Perceptions of corruption"] > upper_bound)]

,year,Overall rank,Country,Region,Score,GDP per Capita,Social support,Health (Life Expectancy),Freedom,Generosity,Perceptions of corruption
0,2015,1,Switzerland,Western Europe,7.587,1.39651,1.34951,0.94143,0.66557,0.29678,0.41978
2,2015,3,Denmark,Western Europe,7.527,1.32548,1.36058,0.87464,0.64938,0.34139,0.48357
3,2015,4,Norway,Western Europe,7.522,1.45900,1.33095,0.88521,0.66973,0.34699,0.36503
4,2015,5,Canada,North America,7.427,1.32629,1.32261,0.90563,0.63297,0.45811,0.32957
5,2015,6,Finland,Western Europe,7.406,1.29025,1.31826,0.88911,0.64169,0.23351,0.41372
...,...,...,...,...,...,...,...,...,...,...,...
633,2019,8,New Zealand,Australia and New Zealand,7.307,1.30300,1.55700,1.02600,0.58500,0.33000,0.38000
639,2019,14,Luxembourg,Western Europe,7.090,1.60900,1.47900,1.01200,0.52600,0.19400,0.31600
641,2019,16,Ireland,Western Europe,7.021,1.49900,1.55300,0.99900,0.51600,0.29800,0.31000
659,2019,34,Singapore,Southeastern Asia,6.262,1.57200,1.46300,1.14100,0.55600,0.27100,0.45300
